File: 0082_remove-duplicates-from-sorted-list-ii.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        if(head == None):
            return head
        checker = head
        hashed_set = set()
        while (checker.next != None):
            if (checker.next.val == checker.val):
                hashed_set.add(checker.val)
            checker = checker.next

        dummy = ListNode(101)
        new_curr = dummy
        curr = head
        while (curr != None):
            if curr.val not in hashed_set:
                new_curr.next = ListNode(curr.val)
                new_curr = new_curr.next
            curr = curr.next
        return dummy.next

File: 0082_remove-duplicates-from-sorted-list-ii.cpp
/**
 * Definition for singly-linked list. */
// struct ListNode
// {
//     int val;
//     ListNode* next;
//     ListNode() : val(0), next(nullptr) {}
//     ListNode(int x) : val(x), next(nullptr) {}
//     ListNode(int x, ListNode* next) : val(x), next(next) {}
// };

class Solution
{
public:
    ListNode* deleteDuplicates(ListNode* head)
    {
        auto dummy = new ListNode{-101, head}, b = dummy, q = dummy, p = head;
        auto r = dummy->val;
        while (p != nullptr)
        {
            if (q->val == p->val)
            {
                r = q->val;
            }
            else
            {
                if (q->val != r)
                {
                    b->next = q;
                    b = q;
                }
                q = p;
            }
            p = p->next;
        }
        if (q->val != r)
        {
            b->next = q;
        }
        else {
            b->next = nullptr;
        }
        return dummy->next;
    }
};
